File: app/src/main/java/com/mvvm/contactlist/repo/CategoryRepository.java
package com.mvvm.contactlist.repo;

import com.mvvm.contactlist.db.dao.CategoryDao;
import com.mvvm.contactlist.db.entities.CategoryEntity;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;

public class CategoryRepository {

    private static CategoryRepository repository;
    private final CategoryDao categoryDao;

    private CategoryRepository(CategoryDao categoryDao) {
        this.categoryDao = categoryDao;
    }

    public static CategoryRepository getInstance(CategoryDao categoryDao) {
        if (repository == null) {
            repository = new CategoryRepository(categoryDao);
        }

        return repository;
    }

    public Flowable<List<CategoryEntity>> getCategory() {
        return categoryDao.getCategory();
    }

    public Completable insert(CategoryEntity entity) {
        return categoryDao.insert(entity);
    }

    public Completable update(CategoryEntity entity) {
        return categoryDao.update(entity);
    }

    public Completable delete(CategoryEntity entity) {
        return categoryDao.delete(entity);
    }
}

File: app/src/main/java/com/mvvm/contactlist/repo/CategoryRepository.kt
package com.mvvm.contactlist.repo

import com.mvvm.contactlist.db.dao.CategoryDao
import com.mvvm.contactlist.db.entities.CategoryEntity
import io.reactivex.Completable
import io.reactivex.Flowable

class CategoryRepository private constructor(private val categoryDao: CategoryDao) {
    val category: Flowable<List<CategoryEntity>>
        get() = categoryDao.category

    fun insert(entity: CategoryEntity): Completable {
        return categoryDao.insert(entity)
    }

    fun update(entity: CategoryEntity): Completable {
        return categoryDao.update(entity)
    }

    fun delete(entity: CategoryEntity): Completable {
        return categoryDao.delete(entity)
    }

    companion object {
        private lateinit var repository: CategoryRepository

        fun getInstance(categoryDao: CategoryDao): CategoryRepository {
            if (!::repository.isInitialized) {
                repository = CategoryRepository(categoryDao)
            }
            return repository
        }
    }
}